src/dnn/convolution.jl

```diff
@@ -0,0 +1,282 @@
+        handle,
+        a_desc=(is_fwd ? b_desc : a_desc),
+        b_desc=(is_fwd ? a_desc : b_desc), conv_desc, c_desc)
+
+    workspace = Workspace(dev, workspace_size)
+    perf_results = find_conv_algo(conv_type;
+        handle, workspace, a, a_desc, b, b_desc, conv_desc, c, c_desc)
+    set_benchmark_cache!(conv_type, conv_args, perf_results)
+
+    perf_results, workspace
+end
+
+function _get_conv_descriptors(args...; padding, stride, dilation, groups)
+    pi32, si32, di32 = map(v -> Int32[reverse(v)...], (padding, stride, dilation))
+    descriptors = TensorDescriptor.(args)
+    cdesc = ConvolutionDescriptor(ndims(args[1]) - 2, pi32, si32, di32, groups)
+    (descriptors..., cdesc, pi32, si32, di32)
+end
+
+function _check_conv_args_dims(K, N)
+    K == N - 2 && return nothing
+    throw(ArgumentError(
+        "Convolution arguments (`padding`, `stride`, `dilation`) must have " *
+        "K = N - 2 dimensions, instead:\n" *
+        "- N=$N (input ndims)\n" *
+        "- K=$K (conv args ndims)\n"))
+end
+
+function convolution!(
+    y::ROCArray{T, N}, ydesc::TensorDescriptor,
+    x::ROCArray{T, N}, xdesc::TensorDescriptor,
+    w::ROCArray{T, N}, wdesc::TensorDescriptor, cdesc::ConvolutionDescriptor;
+    padding, stride, dilation, groups,
+) where {T, N}
+    handle = get_handle()
+    conv_args = ConvolutionArgs(x, w; handle, groups, padding, stride, dilation)
+    perf_results, workspace = find_algorithm(
+        miopenConvFwdAlgorithm_t, handle, conv_args,
+        x, xdesc, w, wdesc, cdesc, y, ydesc)
+
+    # alpha, beta = Ref{T}(one(T)), Ref{T}(zero(T))
+    alpha, beta = Ref{Float32}(one(Float32)), Ref{Float32}(zero(Float32))
+    miopenConvolutionForward(
+        handle, alpha, xdesc.handle, x, wdesc.handle, w, cdesc.handle,
+        perf_results.fwd_algo, beta, ydesc.handle, y,
+        workspace.data.ptr, perf_results.memory) |> check
+    y
+end
+
+function convolution!(
+    y::ROCArray{T, N}, x::ROCArray{T, N}, w::ROCArray{T, N};
+    padding::NTuple{K, Int}, stride::NTuple{K, Int},
+    dilation::NTuple{K, Int}, groups::Int,
+) where {T, N, K}
+    _check_conv_args_dims(K, N)
+    xdesc, wdesc, ydesc, cdesc, pi32, si32, di32 = _get_conv_descriptors(
+        x, w, y; padding, stride, dilation, groups)
+    convolution!(
+        y, ydesc, x, xdesc, w, wdesc, cdesc;
+        padding=pi32, stride=si32, dilation=di32, groups)
+end
+
+function convolution(
+    x::ROCArray{T, N}, w::ROCArray{T, N};
+    padding::NTuple{K, Int}, stride::NTuple{K, Int},
+    dilation::NTuple{K, Int}, groups::Int,
+) where {T, N, K}
+    _check_conv_args_dims(K, N)
+    xdesc, wdesc, cdesc, pi32, si32, di32 = _get_conv_descriptors(
+        x, w; padding, stride, dilation, groups)
+    y = similar(x, T, get_output_size(cdesc, xdesc, wdesc))
+    convolution!(
+        y, TensorDescriptor(y), x, xdesc, w, wdesc, cdesc;
+        padding=pi32, stride=si32, dilation=di32, groups)
+end
+
+function ∇convolution_weight!(
```

Comment:
I'm ok with using Unicode in the names for these functions, but I would also like to add another function like `delta_convolution_weight!` or similar, which forwards to `∇convolution_weight!`, for users who can't easily use Unicode with their keyboards.